File: 6.hashtable/1394.FindLuckyIntegerInAnArray/main.go
package main

// https://leetcode.com/problems/find-lucky-integer-in-an-array/

// hash table
// TC: O(n) SC: O(n)
func findLucky1(arr []int) int {
	freqs := make(map[int]int)

	for _, num := range arr {
		freqs[num]++
	}

	result := -1
	for k, v := range freqs {
		if k == v {
			result = max(result, v)
		}
	}

	return result
}

// counting
// TC: O(n) SC: O(n)
func findLucky2(arr []int) int {
	freqs := make([]int, 501)

	for _, num := range arr {
		freqs[num]++
	}

	result := -1
	for _, num := range arr {
		if freqs[num] == num {
			result = max(result, num)
		}
	}

	return result
}

File: 6.hashtable/1394.FindLuckyIntegerInAnArray/main.py
from typing import List

# https://leetcode.com/problems/find-lucky-integer-in-an-array/

# TC: O(n), SC: O(1)
class Solution:
    def findLucky(self, arr: List[int]) -> int:
        freqs = [0] * 501
        for num in arr:
            freqs[num] += 1

        res = -1
        for num in arr:
            if num == freqs[num]:
                res = max(res, num)
        return res
